common/src/main/java/com/tc/net/protocol/transport/ConnectionHealthCheckerImpl.java

```diff
@@ -44,42 +46,59 @@
 public class ConnectionHealthCheckerImpl implements ConnectionHealthChecker {
 
   private final Logger logger;
-  private final Timer                           monitorThread;
+  private static final ScheduledExecutorService monitorThread = createHealthCheckExecutor();
   private final HealthCheckerMonitorThreadEngine monitorThreadEngine;
 
-  private final SetOnceFlag                      shutdown = new SetOnceFlag();
-  private final SetOnceFlag                      started  = new SetOnceFlag();
+  private Future<?>                      task;
   
   public ConnectionHealthCheckerImpl(HealthCheckerConfig healthCheckerConfig, TCConnectionManager connManager, Supplier<Boolean> reachable) {
     Assert.assertNotNull(healthCheckerConfig);
     Assert.eval(healthCheckerConfig.isHealthCheckerEnabled());
     logger = LoggerFactory.getLogger(ConnectionHealthCheckerImpl.class.getName() + ": "
                                      + healthCheckerConfig.getHealthCheckerName());
-    monitorThread = new Timer(healthCheckerConfig.getHealthCheckerName() + " - HealthCheck-Timer", true);
     monitorThreadEngine = getHealthMonitorThreadEngine(healthCheckerConfig, connManager, reachable, logger);
   }
+  
+  private static ScheduledExecutorService createHealthCheckExecutor() {
+    ScheduledThreadPoolExecutor exec = new ScheduledThreadPoolExecutor(1, (r) -> {
+      ThreadGroup grp = Thread.currentThread().getThreadGroup();
+      String id = "";
+      while (grp instanceof TCThreadGroup) {
+        id = "-" + grp.getName();
+        grp = grp.getParent();
+      }
+      Thread t = new Thread(grp, r, "HealthCheck" + id);
+      t.setDaemon(true);
+      return t;
+    });
+    exec.setKeepAliveTime(5, TimeUnit.SECONDS);
+    exec.allowCoreThreadTimeOut(true);
+    return exec;
+  }
 
   private HealthCheckerMonitorThreadEngine getHealthMonitorThreadEngine(HealthCheckerConfig config,
                                                                           TCConnectionManager connectionManager,
                                                                           Supplier<Boolean> reachable,
                                                                           Logger logger) {
-    return new HealthCheckerMonitorThreadEngine(config, connectionManager, ()-> {
-      if (reachable.get()) {
-        return true;
-      } else {
+    Supplier<Boolean> local = () -> {
+      if (!reachable.get()) {
```

Comment:
If you invert this if condition and then inline the Runnable again the changeset is much smaller.